log4j-api/src/main/java/org/apache/logging/log4j/spi/LoggerRegistry.java

```diff
@@ -43,7 +41,7 @@
 @NullMarked
 public class LoggerRegistry<T extends ExtendedLogger> {
 
-    private final Map<String, Map<MessageFactory, WeakReference<T>>> loggerRefByMessageFactoryByName = new HashMap<>();
+    private final Map<MessageFactory, Map<String, T>> loggerByNameByMessageFactory = new WeakHashMap<>();
```

Comment:
Doesn't this change imply that a `Logger` will never be freed unless the associated `MF` becomes unreachable?

Reply:
The entire `LoggerRegistry` will be freed, when the associated `LoggerContext` becomes unavailable.

The three lines:

```java
     logger = newInstance(this, name, messageFactory);
     loggerRegistry.putIfAbsent(name, effectiveMessageFactory, newLogger); 
     return loggerRegistry.getLogger(name, effectiveMessageFactory); 
```
have been in Log4j Core since 2016. These lines should not return `null`.

Reply:
Why aren't we instead trying to fix the existing code, which allows `Logger`s to be freed when they become unreachable?

Reply:
The current code is fixed:

https://github.com/apache/logging-log4j2/blob/89c9b9f81cb12a6a217788a71832bfbd3b6d07b2/log4j-core/src/main/java/org/apache/logging/log4j/core/LoggerContext.java#L527-L531

The changes in this PR are mainly there to allow `log4j-api` version `2.24.2/2.25.0` to work with any `log4j-core` version `2.24.x`.

Reply:
This PR, instead of fixing _"the immediate problem"_, rewrites the `LoggerRegistry` and reverts the behavior to scope the lifetime of `Logger`s with the associated `MF`'s lifetime. I prefer to preserve the old (yet buggy) behavior that `Logger`s will be GC'ed when they become unreachable. IMO, _the fix_ is two lines of code: adding `Reference.reachabilityFence(logger)` before `return logger` in `putIfAbsent()` and `computeIfAbsent()`. Though `Reference.reachabilityFence()` is introduced in Java 9. We can try imitating it with a usage of the value to force a strong reference (e.g., `logger.log(Level.OFF, "nada")`), but VM can get this inlined/eliminated, and we are back to square one. Anyway, I am looking at this. I will see what else can be done.

Reply:
> Currenlty the `Logger` is GC-ed between the call to `putIfAbsent()` and the call to `getLogger()`. Unless I am mistaken, `reachabilityFence` will not help with that.

No, it will. GC kicks in _after_ the `return logger`, i.e., _outside the scope_ where the reference returned by `WeakReference#get()` is assigned to a variable. Creating a strong reference to `logger` again within this scope will solve the problem – I just don't know of a stable way to this without using the reachability fence. This is in detailed explained [here](https://shipilev.net/jvm/anatomy-quarks/8-local-var-reachability/). You can even verify this behavior by creating a `private static final Set<Logger> LOGGERS` and calling `LOGGERS.add(logger)` just before every `return logger` statement.

> Therefore the refactoring from `Map<String, Map<MessageFactory, WeakReference<T>>>` to `Map<String, Map<MessageFactory, T>>` is required.

_Required_? I am not sure. As I stated, I am checking this. Could you give me a day, please?

> The refactoring from `Map<String, Map<MessageFactory, T>>` to `Map<MessageFactory, Map<String, T>>` is indeed not needed. Do you object to this change?

If this is indeed the only way to fix the problem, I don't object it. Please give me a day.

Reply:
I'm not happy about this change, yet I couldn't come up with a better way in this limited amount of time. I leave the judgement to the rest of the maintainers.